### PR TITLE
refactor: NRF Server & processor

### PR DESCRIPTION
**Pull request by KunLee76**
## Server

- [x] copy server side code
- [x] implement server structure & refactor `api_{SVC}.go`

## Processor
- [x] new `processor/processor.go`
- [x] modify `accesstoken.go` , `nfdiscovery.go` & `nfmanagement.go`

### CODE REVIEW

**Comment by ianchen0119** [comment]
Hi @KunLee76 

Please use `go mod tidy` to renew `go.mod` & `go.sum`.
Besides, please help to fix errors reported by CI workflow.

Thanks!

File: pkg/service/init.go [hunk 1]
```diff
@@ -1,46 +1,66 @@
...
 )
 
 type NrfApp struct {
-	cfg    *factory.Config
-	nrfCtx *nrf_context.NRFContext
+	cfg       *factory.Config
+	nrfCtx    *nrf_context.NRFContext
+	ctx       context.Context
+	cancel    context.CancelFunc
+	proc      *processor.Processor
```
Comment (andy89923) [inline pkg/service/init.go:25]: processor

File: internal/sbi/processor/accesstoken.go [hunk 2]
```diff
@@ -12,49 +12,25 @@ import (
...
 	var expiration int32 = 1000
 	scope := request.Scope
 	tokenType := "Bearer"
 	now := int32(time.Now().Unix())
 
-	errResponse := AccessTokenScopeCheck(request)
-	if errResponse != nil {
-		logger.AccTokenLog.Errorf("AccessTokenScopeCheck error: %v", errResponse.Error)
-		return nil, errResponse
+	AccTokenErr := AccessTokenScopeCheck(request)
```
Comment (andy89923) [inline internal/sbi/processor/accesstoken.go:30]: accTokenErr

File: internal/sbi/api_discovery.go [hunk 3]
```diff
@@ -7,31 +7,47 @@
...
+	auth_err := authorizationCheck(c, "nnrf-disc")
 	if auth_err != nil {
 		c.JSON(http.StatusUnauthorized, gin.H{"error": auth_err.Error()})
 		return
 	}
 
 	req := httpwrapper.NewRequest(c.Request, nil)
 	req.Query = c.Request.URL.Query()
-	httpResponse := producer.HandleNFDiscoveryRequest(req)
+	httpResponse := s.processor.HandleNFDiscoveryRequest(req)
```
Comment (andy89923) [inline internal/sbi/api_discovery.go:50]: s.Processor()

File: internal/sbi/server.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+	Pattern string
+	APIFunc gin.HandlerFunc
+}
+
+type Server struct {
+	nrf
+
+	httpServer *http.Server
+	router     *gin.Engine
+	processor  *processor.Processor
```
Comment (andy89923) [inline internal/sbi/server.go:40]: remove processor

Comment (andy89923) [reply internal/sbi/server.go:40]: Could be retrieved from s.Processor()

File: internal/sbi/api_discovery.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		return
+	}
+
+	// req := httpwrapper.NewRequest(c.Request, nil)
+	// req.Query = c.Request.URL.Query()
+	// httpResponse := s.processor.HandleNFDiscoveryRequest(req)
+	query := c.Request.URL.Query()
+	s.Processor().NFDiscoveryProcedure(c, query)
+
+	//--------------------第2層
```
Comment (ianchen0119) [inline internal/sbi/api_discovery.go:49]: remove unused code

File: internal/sbi/api_accesstoken.go [hunk 6]
```diff
@@ -17,17 +17,38 @@ import (
...
 	"github.com/free5gc/openapi/models"
-	"github.com/free5gc/util/httpwrapper"
 )
 
-// AccessTokenRequest - Access Token Request
-func HTTPAccessTokenRequest(c *gin.Context) {
-	logger.AccTokenLog.Infoln("In HTTPAccessTokenRequest")
+// Index is the index handler.
+func Index(c *gin.Context) {
+	c.String(http.StatusOK, "Hello World!")
```
Comment (tim1207) [inline internal/sbi/api_accesstoken.go:26]: "Hello World!"  replace msg

File: internal/sbi/api_accesstoken.go [hunk 7]
```diff
@@ -77,7 +98,9 @@ func HTTPAccessTokenRequest(c *gin.Context) {
 		}
 	}
 
-	err = c.Bind(&accessTokenReq)
+	if err1 := s.bindData(c, &accessTokenReq); err1 != nil {
```
Comment (tim1207) [inline internal/sbi/api_accesstoken.go:101]: err1 should be err ?

File: internal/sbi/processor/nfmanagement.go [hunk 8]
```diff
@@ -1,14 +1,15 @@
...
+package processor
 
 import (
 	"context"
 	"encoding/json"
 	"fmt"
 	"net/http"
 	"os"
-	"strconv"
+	//"strconv"
```
Comment (tim1207) [inline internal/sbi/processor/nfmanagement.go:9]: remove unused line

File: internal/sbi/api_discovery.go (hunk 3 above)
Comment (KunLee76) [reply internal/sbi/api_discovery.go]: OK

File: internal/sbi/processor/accesstoken.go (hunk 2 above)
Comment (KunLee76) [reply internal/sbi/processor/accesstoken.go]: OK

File: internal/sbi/server.go [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+		AllowMethods: []string{"GET", "POST", "OPTIONS", "PUT", "PATCH", "DELETE"},
+		AllowHeaders: []string{
+			"Origin", "Content-Length", "Content-Type", "User-Agent",
+			"Referrer", "Host", "Token", "X-Requested-With",
+		},
+		ExposeHeaders:    []string{"Content-Length"},
+		AllowCredentials: true,
+		AllowAllOrigins:  true,
+		MaxAge:           CorsConfigMaxAge,
+	}))
```
[the remark this answers is not in the document]
Comment (KunLee76) [reply internal/sbi/server.go]: OK

File: internal/sbi/api_accesstoken.go (hunk 6 above)
Comment (KunLee76) [reply internal/sbi/api_accesstoken.go:26]: OK

File: internal/sbi/api_accesstoken.go (hunk 7 above)
Comment (KunLee76) [reply internal/sbi/api_accesstoken.go:101]: No, the err1 is different from err. 
The err was previously declared and initialized to store the error value returned by the r.ParseForm() (see in line 64)

File: internal/sbi/processor/nfmanagement.go (hunk 8 above)
Comment (KunLee76) [reply internal/sbi/processor/nfmanagement.go:9]: OK

**Comment by andy89923** [comment]
Remember to remove unused code.

File: internal/sbi/server.go [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+		logger.SBILog.Errorf("Bind Request Body error: %v", err)
+		gc.JSON(http.StatusBadRequest,
+			openapi.ProblemDetailsMalformedReqSyntax(err.Error()))
+		return err
+	}
+
+	return nil
+}
+
+func (s *Server) buildAndSendHttpResponse(
```
Comment (andy89923) [inline internal/sbi/server.go:188]: I think we don't use this.

File: internal/sbi/server.go [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+
+	if err != nil {
+		logger.SBILog.Errorln(err)
+		gc.JSON(http.StatusInternalServerError, openapi.ProblemDetailsSystemFailure(err.Error()))
+	} else {
+		gc.Data(rsp.Status, contentType, rspBody)
+	}
+}
+
+func buildHttpResponseHeader(gc *gin.Context, rsp *httpwrapper.Response) {
```
Comment (andy89923) [inline internal/sbi/server.go:221]: I think we don't use this.

File: pkg/factory/config.go [hunk 12]
```diff
@@ -30,6 +30,7 @@ const (
 	NrfSbiDefaultScheme          = "https"
 	NrfNfmResUriPrefix           = "/nnrf-nfm/v1"
 	NrfDiscResUriPrefix          = "/nnrf-disc/v1"
+	NrfAccTokenResUriPrefix      = ""
```
Comment (andy89923) [inline pkg/factory/config.go:33]: ?

**Review by andy89923 — CHANGES_REQUESTED** [review]
Please add the following interface
- pkg/app/app.go

**Pull request closed without merging**